lib/Transforms/Scalar/DxilLoopUnroll.cpp

```diff
@@ -0,0 +1,816 @@
+          PHINode *PN = dyn_cast<PHINode>(&I);
+          if (!PN)
+            break;
+          Value *OldIncoming = PN->getIncomingValueForBlock(BB);
+          Value *NewIncoming = OldIncoming;
+          ValueToValueMapTy::iterator Itor = CurIteration.VarMap.find(OldIncoming);
+          if (Itor != CurIteration.VarMap.end())
+            NewIncoming = Itor->second;
+          PN->addIncoming(NewIncoming, ClonedBB);
+        }
+      }
+    }
+
+    // Remap the instructions inside of cloned blocks.
+    for (BasicBlock *BB : CurIteration.Body) {
+      for (Instruction &I : *BB) {
+        ::RemapInstruction(&I, CurIteration.VarMap);
+      }
+    }
+
+    // Remap branch instruction successors.
+    for (BasicBlock *ClonedBB : CurIteration.Body) {
+      BranchInst *BI = dyn_cast<BranchInst>(ClonedBB->getTerminator());
+      if (!BI)
+        continue;
+
+      for (unsigned j = 0, NumSucc = BI->getNumSuccessors(); j < NumSucc; j++) {
+        BasicBlock *OldSucc = BI->getSuccessor(j);
+        ValueToValueMapTy::iterator Iter = CurIteration.VarMap.find(OldSucc);
+        if (Iter != CurIteration.VarMap.end()) {
+          BI->setSuccessor(j, cast<BasicBlock>(Iter->second));
+        }
+      }
+    }
+
+    // If this is the first block
+    if (!PrevIteration) {
+      // Replace the phi nodes in the clone block with the values coming
+      // from outside of the loop
+      for (PHINode *PN : PHIs) {
+        PHINode *ClonedPN = cast<PHINode>(CurIteration.VarMap[PN]);
+        Value *ReplacementVal = ClonedPN->getIncomingValueForBlock(Predecessor);
+        ClonedPN->replaceAllUsesWith(ReplacementVal);
+        ClonedPN->eraseFromParent();
+        CurIteration.VarMap[PN] = ReplacementVal;
+      }
+    }
+    else {
+      for (PHINode *PN : PHIs) {
+        PHINode *ClonedPN = cast<PHINode>(CurIteration.VarMap[PN]);
+        Value *ReplacementVal = PrevIteration->VarMap[PN->getIncomingValueForBlock(Latch)];
+        ClonedPN->replaceAllUsesWith(ReplacementVal);
+        ClonedPN->eraseFromParent();
+        CurIteration.VarMap[PN] = ReplacementVal;
+      }
+
+      // Make the latch of the previous iteration branch to the header
+      // of this new iteration.
+      if (BranchInst *BI = dyn_cast<BranchInst>(PrevIteration->Latch->getTerminator())) {
+        for (unsigned i = 0; i < BI->getNumSuccessors(); i++) {
+          if (BI->getSuccessor(i) == PrevIteration->Header) {
+            BI->setSuccessor(i, CurIteration.Header);
+            break;
+          }
+        }
+      }
+    }
+
+    // Simplify instructions in the cloned blocks to create
+    // constant exit conditions.
+    for (BasicBlock *ClonedBB : CurIteration.Body)
+      SimplifyInstructionsInBlock_NoDelete(ClonedBB, NULL);
+
+    // Check exit condition to see if we fully unrolled the loop
+    if (BranchInst *BI = dyn_cast<BranchInst>(CurIteration.Latch->getTerminator())) {
+      bool Cond = false;
+      if (!GetConstantI1(BI->getCondition(), &Cond))
```

Comment:
Actually, couldn't this be non-const until it exits the loop at an upper bound if the condition is something like `(i < var && i < 4)`?  If so, we would want to continue unrolling until we either finish the loop or hit MaxIterations, right?